# Wrangle Pisa 2012
#### By Gabriela Sikora

Throughout this notebook, I will be assessing and cleaning data from the PISA 2012 dataset. The purpose of this wrangling is to have a usable dataframe for the complementary exploration notebook, wherein I will be exploring the cleaned dataframe and drawing conclusions from the plots.


### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [3]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [4]:
# Quick overview of the entire dataframe
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [5]:
# Check how many entries and types of data
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [6]:
# Look through the datatypes
pisa.dtypes[400:]

ST23Q08        object
ST24Q01        object
ST24Q02        object
ST24Q03        object
CLCUSE1        object
CLCUSE301       int64
CLCUSE302       int64
DEFFORT         int64
QUESTID        object
BOOKID         object
EASY           object
AGE           float64
GRADE         float64
PROGN          object
ANXMAT        float64
ATSCHL        float64
ATTLNACT      float64
BELONG        float64
BFMJ2         float64
BMMJ1         float64
CLSMAN        float64
COBN_F         object
COBN_M         object
COBN_S         object
COGACT        float64
CULTDIST      float64
CULTPOS       float64
DISCLIMA      float64
ENTUSE        float64
ESCS          float64
               ...   
W_FSTR55      float64
W_FSTR56      float64
W_FSTR57      float64
W_FSTR58      float64
W_FSTR59      float64
W_FSTR60      float64
W_FSTR61      float64
W_FSTR62      float64
W_FSTR63      float64
W_FSTR64      float64
W_FSTR65      float64
W_FSTR66      float64
W_FSTR67      float64
W_FSTR68      float64
W_FSTR69  

In [7]:
# Check dictionary for names of variables
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [8]:
# Check for null values in relevant columns
columns = ['CNT', 'STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

CNT 0
STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [9]:
# See countries included in dataframe
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [10]:
# Check if there are any duplicates
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
3. Make the variable names readable since they are codes that can only be read with the dictionary

4. There are multiple columns for single variables related to father's education
5. There are multiple columns for single variables related to mother's education

6. Missing values in relevant columns

7. Create a total for Out-of-School Study Time and Learning time (minutes per week)
8. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [11]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### 1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed


There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [12]:
# Create new dataframe with only the selected columns
pisa_clean = pisa_clean[['CNT', 'STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [13]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['CNT',
 'STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### 2. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [14]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score']) / 3

In [15]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [16]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())
print('-'*22)

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
----------------------
count    485490.000000
mean        472.488764
std          96.036271
min          77.114593
25%         403.992595
50%         472.046460
75%         541.455700
max         826.592027
Name: Average Total Score, dtype: float64


In [17]:
# Test to see if the data from the new variables display correctly
pisa_clean.sample(10)

CNT  STIDSTD ST04Q01                ST13Q01  \
231572                   Ireland     4562    Male      <ISCED level 3A>    
463431                   Tunisia     2449    Male       <ISCED level 2>    
124556                Costa Rica     1385  Female       <ISCED level 2>    
111635                     Chile     4393  Female      <ISCED level 3A>    
335290                    Mexico    19588    Male      <ISCED level 3A>    
374017                      Peru     1131    Male                    NaN   
185670                    France     1167    Male  <ISCED level 3B, 3C>    
407705             Florida (USA)      551  Female      <ISCED level 3A>    
94247                     Canada    19778  Female      <ISCED level 3A>    
406269  Perm(Russian Federation)      876  Female      <ISCED level 3A>    

       ST14Q01 ST14Q02 ST14Q03 ST14Q04                ST17Q01 ST18Q01  ...  \
231572     NaN     NaN     NaN     NaN       <ISCED level 2>      NaN  ...   
463431      No      No      No      No       <ISCED level 1>       No  ...   
124556     NaN     NaN     NaN     Yes                    NaN      No  ...   
111635      No      No     Yes      No      <ISCED level 3A>      Yes  ...   
335290      No      No      No     NaN                    NaN     NaN  ...   
374017     NaN     Yes     NaN     NaN       <ISCED level 2>      NaN  ...   
185670      No      No     Yes     NaN  <ISCED level 3B, 3C>       No  ...   
407705     NaN     NaN     NaN     Yes       <ISCED level 2>      NaN  ...   
94247       No     Yes      No      No      <ISCED level 3A>       No  ...   
406269      No     Yes     Yes      No                    NaN     NaN  ...   

       ST57Q03 ST57Q04 ST57Q05  MMINS  LMINS  SMINS  Average Math Score  \
231572     NaN     NaN     NaN    NaN    NaN    NaN           458.49040   
463431     NaN     NaN     NaN    NaN    NaN    NaN           339.31296   
124556     0.0     0.0     1.0  200.0  200.0  200.0           361.12324   
111635     NaN     NaN     NaN    NaN    NaN    NaN           527.03692   
335290     NaN     NaN     NaN    NaN    NaN    NaN           504.68140   
374017     2.0     NaN     NaN    NaN    NaN    NaN           231.19644   
185670     3.0     3.0     3.0  180.0  240.0  240.0           548.53560   
407705     NaN     NaN     NaN    NaN    NaN    NaN           459.65882   
94247      0.0     0.0     0.0  250.0  250.0  250.0           604.61908   
406269     NaN     NaN     3.0  200.0   80.0  200.0           363.84950   

        Average Reading Score  Average Science Score  Average Total Score  
231572              440.97132              463.29726           454.252993  
463431              327.17534              391.49570           352.661333  
124556              472.30076              393.08094           408.834980  
111635              507.09154              520.36554           518.164667  
335290              493.33834              491.73816           496.585967  
374017              217.79004              265.51660           238.167693  
185670              562.54624              538.08282           549.721553  
407705              461.02154              457.32936           459.336573  
94247               583.02758              624.52442           604.057027  
406269              367.92838              404.73704           378.838307  

[10 rows x 25 columns]

### 3. Make the variable names readable since they are codes that can only be read with the dictionary

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [18]:
# Rename columns
pisa_clean.rename({'CNT': 'Country',
                   'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning Time - Mathematics',
                   'LMINS': 'Learning Time - Test Language',
                   'SMINS': 'Learning Time - Science'}, axis='columns', inplace=True)

In [19]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Country',
 'Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

### 4. There are multiple columns for single variables related to father's education

Values for the father's education are distributed among 5 columns. One column holds '<ISCED level 1>', '<ISCED level 2>', '<ISCED level 3A>', '<ISCED level 3B, 3C>', and 'He did not complete <ISCED level 1>'. We would rather have one level for ISCED level 3, and change 'He did not complete <ISCED level 1>' to level 0, in accordance to this classification: http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

Also, since the data for this variable has multiple pieces of data in one column, and then another 4 columns hold the rest of the data, we will have to expand the first column by creating dummies for each data to have a column, and then melt all the columns into one. 

In [20]:
# Begin with a copy
father_edu = pisa_clean.copy()

In [21]:
# Split column up with dummies
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [22]:
# Rename the columns for consistency and to remove subcategorization of levels
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [23]:
# Change datatypes from ints into strings for dummy columns
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [24]:
# Change values from '1's and 'Yes's to the corresponding level to prepare for the melt
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [25]:
# Test to see if the columns are all correct before melt
father_edu.sample(10)

Country  Student ID  Gender  \
176901               Finland        1227    Male   
5897    United Arab Emirates        1155  Female   
59292                 Brazil        4027    Male   
362528           Netherlands        3079  Female   
466201                Turkey         812    Male   
210407       Hong Kong-China        3507    Male   
367760                Norway        3851    Male   
19476              Argentina        3234    Male   
64945                 Brazil        9680    Male   
82567                 Canada        8098  Female   

                   Mother - Highest Schooling  \
176901                  <ISCED level 3B, 3C>    
5897                        <ISCED level 3A>    
59292   She did not complete <ISCED level 1>    
362528                      <ISCED level 3A>    
466201                       <ISCED level 1>    
210407                       <ISCED level 1>    
367760                  <ISCED level 3B, 3C>    
19476                       <ISCED level 3A>    
64945                   <ISCED level 3B, 3C>    
82567                       <ISCED level 3A>    

       Mother Qualifications - ISCED level 6  \
176901                                    No   
5897                                     NaN   
59292                                     No   
362528                                    No   
466201                                    No   
210407                                    No   
367760                                    No   
19476                                    Yes   
64945                                    Yes   
82567                                     No   

       Mother Qualifications - ISCED level 5A  \
176901                                    Yes   
5897                                      NaN   
59292                                      No   
362528                                     No   
466201                                     No   
210407                                     No   
367760                                     No   
19476                                      No   
64945                                     NaN   
82567                                      No   

       Mother Qualifications - ISCED level 5B  \
176901                                     No   
5897                                      NaN   
59292                                      No   
362528                                     No   
466201                                     No   
210407                                     No   
367760                                    Yes   
19476                                      No   
64945                                     NaN   
82567                                     Yes   

       Mother Qualifications - ISCED level 4 Father - Highest Schooling  \
176901                                    No          <ISCED level 3A>    
5897                                     NaN           <ISCED level 2>    
59292                                     No          <ISCED level 3A>    
362528                                   Yes                        NaN   
466201                                   NaN           <ISCED level 1>    
210407                                    No      <ISCED level 3B, 3C>    
367760                                   NaN      <ISCED level 3B, 3C>    
19476                                    NaN          <ISCED level 3A>    
64945                                    NaN      <ISCED level 3B, 3C>    
82567                                     No          <ISCED level 3A>    

        <ISCED level 6>  ... Learning Time - Science Average Math Score  \
176901               No  ...                   120.0          582.88674   
5897                NaN  ...                     NaN          397.26592   
59292                No  ...                     NaN          481.46908   
362528               No  ...                     0.0          458.17886   
466201               No  ...                    90.0          330.58888   
210407               No

In [26]:
# Combine all columns into one variable column
father_edu = pd.melt(father_edu, 
                     id_vars=['Country','Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [27]:
# Check to see if the columns from the melt were correctly created
father_edu.sample(10)

Country  Student ID  Gender Mother - Highest Schooling  \
851284      Norway        1885  Female           <ISCED level 2>    
4135443      Italy       10934    Male      <ISCED level 3B, 3C>    
2697353      Italy       29314  Female          <ISCED level 3A>    
557241      Brazil       16486    Male           <ISCED level 1>    
4601650    Iceland         214  Female          <ISCED level 3A>    
4215590     Mexico       15968    Male           <ISCED level 1>    
152999       Spain        7417    Male           <ISCED level 1>    
2560977    Germany         427    Male           <ISCED level 2>    
1883010  Singapore        3765    Male          <ISCED level 3A>    
1694894     Israel        2890  Female          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
851284                                     No   
4135443                                    No   
2697353                                    No   
557241                                     No   
4601650                                    No   
4215590                                    No   
152999                                     No   
2560977                                   NaN   
1883010                                   NaN   
1694894                                   Yes   

        Mother Qualifications - ISCED level 5A  \
851284                                      No   
4135443                                     No   
2697353                                     No   
557241                                      No   
4601650                                     No   
4215590                                     No   
152999                                      No   
2560977                                    NaN   
1883010                                    Yes   
1694894                                    Yes   

        Mother Qualifications - ISCED level 5B  \
851284                                      No   
4135443                                    NaN   
2697353                                     No   
557241                                      No   
4601650                                     No   
4215590                                     No   
152999                                      No   
2560977                                    Yes   
1883010                                    NaN   
1694894                                     No   

        Mother Qualifications - ISCED level 4  \
851284                                     No   
4135443                                   NaN   
2697353                                    No   
557241                                     No   
4601650                                    No   
4215590                                   NaN   
152999                                     No   
2560977                                   NaN   
1883010                                   NaN   
1694894                                    No   

         Out-of-School Study Time - Homework  \
851284                                   0.0   
4135443                                  NaN   
2697353                                  NaN   
557241                                   NaN   
4601650                                  3.0   
4215590                                  NaN   
152999                                   1.0   
2560977                                  NaN   
1883010                                 14.0   
1694894                                  4.0   

         Out-of-School Study Time - Guided Homework  ...  \
851284                                          0.0  ...   
4135443                                         NaN  ...   
2697353                                         NaN  ...   
557241                                          NaN  ...   
4601650                                         2.0  ...   
4215590                                         NaN  ...   
152999                                          NaN  ...   
2560977                                         NaN  ...   


In [28]:
# Remove all cases from melt where the education of the father is either '0' or 'No'
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [29]:
# Check to see if 0's and No's have been removed
father_edu.sample(70)

Country  Student ID  Gender  \
1173705                    Greece         950  Female   
3809798       Massachusetts (USA)         621  Female   
2373457                    Serbia        3176    Male   
3507385                     Chile        1713  Female   
78268                      Canada        3799  Female   
94007                      Canada       19538  Female   
2177297                   Iceland        3311    Male   
162860                      Spain       17278  Female   
3550026                     Spain        6014  Female   
3280442                    Norway        3593    Male   
972738                    Albania        1759  Female   
214810                    Croatia        3240  Female   
103000                Switzerland        6987    Male   
307956                     Latvia        1895  Female   
3815647                   Romania        4747  Female   
1801116                    Mexico       28944  Female   
4104228                   Hungary        3730    Male   
2261381                    Mexico        3719    Male   
475710   United States of America         158    Male   
322475                     Mexico        6773    Male   
4249266                    Norway        1437    Male   
2758015                    Mexico       14863  Female   
2142095            United Kingdom       11019    Male   
451471             Chinese Taipei        3141  Female   
4191580                    Latvia        1599    Male   
504046                  Argentina        2314  Female   
29493                   Australia        7343    Male   
1754226                 Lithuania        1571    Male   
1094465                Costa Rica         314  Female   
3414561      United Arab Emirates       11389  Female   
...                           ...         ...     ...   
87154                      Canada       12685    Male   
2237185                     Korea        4366  Female   
907768         Russian Federation        4734  Female   
311016                Macao-China         649    Male   
291430                      Korea         571    Male   
2260151                    Mexico        2489    Male   
3411914      United Arab Emirates        8742    Male   
297980                  Lithuania        1795  Female   
342710                     Mexico       27008    Male   
599786                   Colombia         198    Male   
147661                      Spain        2079  Female   
1438607                    Turkey        2238    Male   
4233466                Montenegro          38    Male   
600858                   Colombia        1270  Female   
1858310            China-Shanghai        1624  Female   
4715411                    Mexico       30299    Male   
863615                       Peru        5239    Male   
4065005                   Finland        5411  Female   
420873         Russian Federation        3329    Male   
1974154                 Australia       10044    Male   
885993             China-Shanghai         287  Female   
1654007            United Kingdom        8421    Male   
394212                      Qatar        4962    Male   
305859                 Luxembourg        5056  Female   
2695901                     Italy       27862    Male   
3455255                    Brazil        1560    Male   
863131                       Peru        4755    Male   
3086004                   Estonia        2169    Male   
600968                   Colombia        1380  Female   
4024370                   Denmark        2349  Female   

                    Mother - Highest Schooling  \
1173705                      <ISCED level 3A>    
3809798                      <ISCED level 3A>    
2373457                  <ISCED level 3B, 3C>    
3507385                      <ISCED level 3A>    
78268                        <ISCED level 3A>    
94007                        <ISCED level 3A>    
2177297                      <ISCED level 3A>    
162860                        <ISCED level 2>    
3550026                  <ISCED level 3B,

In [30]:
# Remove 'Father - Highest Schooling' since it is redundant
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [31]:
# Check to see if correct values remain
father_edu['Education - Father'].value_counts()

<ISCED level 3>    306459
<ISCED level 5>    185130
<ISCED level 2>     84329
<ISCED level 4>     77265
<ISCED level 1>     35938
<ISCED level 6>     29589
<ISCED level 0>     16535
Name: Education - Father, dtype: int64

In [32]:
# Completed_edu is no longer needed since it duplicates the information in 'Education - Father'
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [33]:
# Ensure the column has been dropped
father_edu.sample(10)

Country  Student ID  Gender  \
1985402               Belgium        2056    Male   
3717456                Mexico        3324  Female   
1703921                 Italy        6862  Female   
1295789                Mexico        9107  Female   
4057873               Estonia        3058    Male   
903015                Romania        5055  Female   
1759712            Luxembourg        2439    Male   
2195554                 Italy       13005  Female   
982679   United Arab Emirates        6957  Female   
2316932                  Peru        2086  Female   

                    Mother - Highest Schooling  \
1985402                  <ISCED level 3B, 3C>    
3717456                      <ISCED level 3A>    
1703921                  <ISCED level 3B, 3C>    
1295789                      <ISCED level 3A>    
4057873                      <ISCED level 3A>    
903015                        <ISCED level 2>    
1759712                      <ISCED level 3A>    
2195554                  <ISCED level 3B, 3C>    
982679                       <ISCED level 3A>    
2316932  She did not complete <ISCED level 1>    

        Mother Qualifications - ISCED level 6  \
1985402                                   NaN   
3717456                                    No   
1703921                                   NaN   
1295789                                   NaN   
4057873                                    No   
903015                                    NaN   
1759712                                   NaN   
2195554                                   NaN   
982679                                    NaN   
2316932                                    No   

        Mother Qualifications - ISCED level 5A  \
1985402                                    NaN   
3717456                                     No   
1703921                                    Yes   
1295789                                    NaN   
4057873                                     No   
903015                                     NaN   
1759712                                    NaN   
2195554                                    NaN   
982679                                     Yes   
2316932                                     No   

        Mother Qualifications - ISCED level 5B  \
1985402                                    Yes   
3717456                                     No   
1703921                                    NaN   
1295789                                    Yes   
4057873                                    Yes   
903015                                     NaN   
1759712                                    Yes   
2195554                                    NaN   
982679                                     NaN   
2316932                                     No   

        Mother Qualifications - ISCED level 4  \
1985402                                   NaN   
3717456                                   NaN   
1703921                                   NaN   
1295789                                   NaN   
4057873                                   Yes   
903015                                    NaN   
1759712                                   NaN   
2195554                                   NaN   
982679                                    NaN   
2316932                                    No   

         Out-of-School Study Time - Homework  \
1985402                                  NaN   
3717456                                  4.0   
1703921                                  2.0   
1295789                                 10.0   
4057873                                  NaN   
903015                                   2.0   
1759712                                  3.0   
2195554                                  NaN   
982679                                  10.0   
2316932                                  4.0   

         Out-of-School Study Time - Guided Homework  ...  \
1985402                                         NaN  ...   
3717456                                         1.0  ...

In [34]:
# Check if there are any duplicated values remaining
father_edu['Student ID'].duplicated().sum()

1229261

In [35]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
# Change datatypes into category, and put into order from lowest to highest
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')
father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['<ISCED level 6>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 0>'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)

# Remove duplicates and keep last, which is the highest level of education for the student's father
father_edu_clean = father_edu.drop_duplicates(subset=['Student ID'], keep='first')

In [36]:
# Ensure there are no duplicates
father_edu_clean['Student ID'].duplicated().sum()

0

In [37]:
# Check the size of the dataframe - 
print(pisa['STIDSTD'].nunique())
print(father_edu_clean.shape)

33806
(33806, 21)


In [38]:
# Final check of clean dataframe
father_edu_clean.sample(100)

Country  Student ID  Gender  \
1725410                     Italy       28351    Male   
898862                    Romania         902    Male   
3261311                    Mexico       32669  Female   
1307221                    Mexico       20539    Male   
571475                     Canada       11516  Female   
1225752                     Italy       14183  Female   
778738                      Korea        2389  Female   
962241   United States of America        1199    Male   
1134969                     Spain       18407  Female   
1004443                 Australia       11313    Male   
894785          Connecticut (USA)         245  Female   
571713                     Canada       11754  Female   
2696779                     Italy       28740    Male   
880099                      Qatar        5359  Female   
755119                      Italy       29040    Male   
844790                   Malaysia        5048    Male   
646067                      Spain       14995    Male   
1652734            United Kingdom        7148    Male   
3184197                     Italy       30668  Female   
650292                      Spain       19220  Female   
1248955                    Jordan        6313  Female   
682989             United Kingdom        8383    Male   
588868                Switzerland        7365    Male   
1528274                    Brazil       16539  Female   
4228731                    Mexico       29109  Female   
1370898                     Qatar       10668  Female   
496937       United Arab Emirates        6705    Male   
1225380                     Italy       13811    Male   
3471940                    Brazil       18245  Female   
819311                     Mexico       18119  Female   
...                           ...         ...     ...   
3004651                    Canada       17242    Male   
1307415                    Mexico       20733  Female   
2028463                    Canada       12034    Male   
3082977                     Spain       24455    Male   
883590                      Qatar        8850    Male   
4151093                     Italy       26584  Female   
1543850                    Canada       12911  Female   
4718173                    Mexico       33061  Female   
1468694      United Arab Emirates        7482    Male   
1490833                 Australia       12213    Male   
746548                      Italy       20469  Female   
3472373                    Brazil       18678  Female   
603566                   Colombia        3978    Male   
1544055                    Canada       13116  Female   
719405                    Iceland        1889  Female   
3260562                    Mexico       31920  Female   
1490322                 Australia       11702  Female   
2035744                    Canada       19315  Female   
1640076                   Finland        7932    Male   
1134487                     Spain       17925  Female   
1315398                    Mexico       28716    Male   
1624090                     Spain       22038    Male   
1240909                     Italy       29340  Female   
1232517                     Italy       20948    Male   
1725638                     Italy       28579    Male   
3179176                     Italy       25647  Female   
1548994                    Canada       18055    Male   
1790675                    Mexico       18503    Male   
3558415                     Spain       14403    Male   
1794796                    Mexico       22624  Female   

                    Mother - Highest Schooling  \
1725410                  <ISCED level 3B, 3C>    
898862                                     NaN   
3261311                       <ISCED level 2>    
1307221                      <ISCED level 3A>    
571475                       <ISCED level 3A>    
1225752                      <ISCED level 3A>    
778738                       <ISCED level 3A>    
962241                       <ISCED level 3A>    
1134969                      <ISCED level

### 5. There are multiple columns for single variables related to mother's education

The exact same process for the mother's education is followed as for the father's education. 

In [39]:
# Build mother's education dataframe off of the existing father education dataframe
mother_edu = father_edu_clean.copy()

In [40]:
# Check to visually assess
mother_edu.head()

Country  Student ID  Gender Mother - Highest Schooling  \
959699   Uruguay        3972    Male           <ISCED level 2>    
899695   Romania        1735  Female      <ISCED level 3B, 3C>    
602994  Colombia        3406  Female                        NaN   
884540     Qatar        9800  Female          <ISCED level 3A>    
649234     Spain       18162    Male           <ISCED level 2>    

       Mother Qualifications - ISCED level 6  \
959699                                   NaN   
899695                                   Yes   
602994                                   Yes   
884540                                   Yes   
649234                                    No   

       Mother Qualifications - ISCED level 5A  \
959699                                    NaN   
899695                                    Yes   
602994                                    NaN   
884540                                    Yes   
649234                                     No   

       Mother Qualifications - ISCED level 5B  \
959699                                    NaN   
899695                                    Yes   
602994                                    NaN   
884540                                    Yes   
649234                                     No   

       Mother Qualifications - ISCED level 4  \
959699                                   NaN   
899695                                   Yes   
602994                                   NaN   
884540                                    No   
649234                                    No   

        Out-of-School Study Time - Homework  \
959699                                  NaN   
899695                                  NaN   
602994                                  5.0   
884540                                  NaN   
649234                                  8.0   

        Out-of-School Study Time - Guided Homework  ...  \
959699                                         NaN  ...   
899695                                         NaN  ...   
602994                                         1.0  ...   
884540                                         NaN  ...   
649234                                         0.0  ...   

        Out-of-School Study Time - Commercial Company  \
959699                                            NaN   
899695                                            NaN   
602994                                            0.0   
884540                                            NaN   
649234                                            0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
959699                                     NaN                          NaN   
899695                                     NaN                          NaN   
602994                                     2.0                        270.0   
884540                                     NaN                          NaN   
649234                                     2.0                        220.0   

        Learning Time - Test Language  Learning Time - Science  \
959699                            NaN                      NaN   
899695                            NaN                      NaN   
602994                          225.0                    225.0   
884540                            NaN                      NaN   
649234                          220.0                    165.0   

        Average Math Score  Average Reading Score  Average Science Score  \
959699           337.75510              363.74404              379.46662   
899695           388.61972              471.26814              397.92986   
602994           393.60492              416.54018              365.19954   
884540           451.47996              461.18040              439.89184   
649234           493.77630              461.58134              465.34874   

        Average Total Score  Education - Father  
959699           360.321920     <ISCED level 6>  
899695           419.272573 

In [41]:
# Compare original dataframe to father dataframe
print(pisa_clean['Mother - Highest Schooling'].value_counts())
print('-'*47)
print(mother_edu['Mother - Highest Schooling'].value_counts())

<ISCED level 3A>                         236993
<ISCED level 3B, 3C>                      83048
<ISCED level 2>                           82614
<ISCED level 1>                           36556
She did not complete <ISCED level 1>      18768
Name: Mother - Highest Schooling, dtype: int64
-----------------------------------------------
<ISCED level 3A>                         20389
<ISCED level 2>                           5160
<ISCED level 3B, 3C>                      5114
<ISCED level 1>                           1347
She did not complete <ISCED level 1>       635
Name: Mother - Highest Schooling, dtype: int64


In [42]:
# Split column up with dummies
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [43]:
# Rename the columns for consistency and to remove subcategorization of levels
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [44]:
# Change datatypes from ints into strings for dummy columns
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [45]:
# Change values from '1's and 'Yes's to the corresponding level to prepare for the melt
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [46]:
# Test to see if the columns are all correct before melt
mother_edu.sample(10)

Country  Student ID  Gender Mother - Highest Schooling  \
3260822                Mexico       32180    Male           <ISCED level 2>    
1797432                Mexico       25260    Male           <ISCED level 2>    
627448                Denmark        3857  Female          <ISCED level 3A>    
2776858                Mexico       33706  Female           <ISCED level 2>    
1313259                Mexico       26577  Female           <ISCED level 2>    
885423                  Qatar       10683    Male          <ISCED level 3A>    
627198                Denmark        3607  Female          <ISCED level 3A>    
686001         United Kingdom       11395    Male          <ISCED level 3A>    
1133650                 Spain       17088    Male      <ISCED level 3B, 3C>    
1472329  United Arab Emirates       11117  Female          <ISCED level 3A>    

        <ISCED level 6>  <ISCED level 5>  <ISCED level 5>  <ISCED level 4>  \
3260822              No               No               No              NaN   
1797432              No               No  <ISCED level 5>              NaN   
627448               No               No  <ISCED level 5>               No   
2776858             NaN              NaN              NaN              NaN   
1313259              No  <ISCED level 5>               No              NaN   
885423               No               No               No  <ISCED level 4>   
627198               No               No  <ISCED level 5>               No   
686001               No  <ISCED level 5>               No  <ISCED level 4>   
1133650              No               No  <ISCED level 5>               No   
1472329              No               No               No  <ISCED level 4>   

         Out-of-School Study Time - Homework  \
3260822                                  NaN   
1797432                                 10.0   
627448                                   5.0   
2776858                                  2.0   
1313259                                  NaN   
885423                                   NaN   
627198                                   NaN   
686001                                   NaN   
1133650                                  NaN   
1472329                                  2.0   

         Out-of-School Study Time - Guided Homework  ...  Average Math Score  \
3260822                                        24.0  ...           479.98908   
1797432                                         5.0  ...           397.73328   
627448                                          0.0  ...           543.16094   
2776858                                         0.0  ...           427.95606   
1313259                                         0.0  ...           349.04968   
885423                                          NaN  ...           429.04660   
627198                                          NaN  ...           640.52812   
686001                                          NaN  ...           402.79640   
1133650                                         NaN  ...           505.14880   
1472329                                         2.0  ...           310.95966   

         Average Reading Score  Average Science Score  Average Total Score  \
3260822              429.34316              383.56954           430.967260   
1797432              416.03084              398.02312           403.929080   
627448               577.54682              569.69416           563.467307   
2776858              450.69550              399.60836           426.086640   
1313259              383.49686              335.54644           356.030993   
885423               410.25684              404.64380           414.649080   
627198               656.02468              642.80120           646.451333   
686001               372.88620              420.40282           398.695140   
1133650              522.44898              473.08840           500.228727   
1472329              346.16430              370.51474           342.546233   

         Educ

In [47]:
# Combine all columns into one variable column
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Country','Student ID', 'Gender',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [48]:
# Check to see if the columns from the melt were correctly created
mother_edu.sample(10)

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
307408       Qatar        4056  Female                                  NaN   
87123       Mexico       32450  Female                                  6.0   
69524   Luxembourg        3108  Female                                  2.0   
162050       Spain       23272  Female                                  2.0   
99335        Italy       30429    Male                                  1.0   
15435       Mexico       33165    Male                                  1.0   
135758       Qatar       10451  Female                                  3.0   
158391       Spain       23219  Female                                  2.0   
172358  Montenegro        3220    Male                                 24.0   
146727      Canada       14902    Male                                 10.0   

        Out-of-School Study Time - Guided Homework  \
307408                                         0.0   
87123                                          2.0   
69524                                          0.0   
162050                                         1.0   
99335                                          0.0   
15435                                          NaN   
135758                                         2.0   
158391                                         0.0   
172358                                        24.0   
146727                                         0.0   

        Out-of-School Study Time - Personal Tutor  \
307408                                        0.0   
87123                                         NaN   
69524                                         0.0   
162050                                        1.0   
99335                                         0.0   
15435                                         1.0   
135758                                        2.0   
158391                                        0.0   
172358                                       24.0   
146727                                        0.0   

        Out-of-School Study Time - Commercial Company  \
307408                                            0.0   
87123                                             NaN   
69524                                             0.0   
162050                                            0.0   
99335                                             0.0   
15435                                             0.0   
135758                                            5.0   
158391                                            2.0   
172358                                           24.0   
146727                                            0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
307408                                     0.0                          NaN   
87123                                      NaN                        250.0   
69524                                      0.0                        200.0   
162050                                     0.0                        200.0   
99335                                      0.0                        200.0   
15435                                      NaN                        200.0   
135758                                     NaN                        180.0   
158391                                     0.0                        135.0   
172358                                    24.0                        135.0   
146727                                     1.0                        375.0   

        Learning Time - Test Language  Learning Time - Science  \
307408                            NaN                      NaN   
87123                           250.0                    250.0   
69524                           250.0                    200.0   
162050                          200.0                      NaN   
99335                           300.0                    100.0   
15435                           225.0                    200.0   
13

In [49]:
# Remove all cases from melt where the education of the mother is either '0' or 'No'
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [50]:
# Check to see if 0's and No's have been removed
mother_edu.sample(70)

Country  Student ID  Gender  \
236087                    Mexico       31879  Female   
247391            Chinese Taipei        5907  Female   
7396                       Spain       14844  Female   
8882    United States of America        3572  Female   
56942                      Spain       23135    Male   
201906                    Mexico       33326  Female   
203488               Netherlands        4076    Male   
39333                     Brazil        8319    Male   
104173                    Norway        1469  Female   
23532                      Spain       18618  Female   
21046                      Spain       16274  Female   
137003                   Romania        2631    Male   
39752                     Mexico        8864    Male   
234484                     Italy       29929    Male   
122789                    Mexico       20672  Female   
292147                    Mexico       33306    Male   
4751                   Australia       12909    Male   
141296                    Mexico       12952  Female   
264238                     Italy       21790  Female   
120508                    Mexico       23966    Male   
71970                     Norway         663  Female   
136517                    Canada         396  Female   
254201                     Spain       24154    Male   
20901                      Spain       13938    Male   
239317                   Romania        4624  Female   
110018                  Malaysia        1226  Female   
9915              United Kingdom        2896  Female   
253056                    Canada       19185  Female   
42056                      Spain       11635    Male   
234984                     Italy       25915  Female   
...                          ...         ...     ...   
119935                     Spain       21207  Female   
138970                     Korea        2671  Female   
111879                    Greece        2828    Male   
237236                  Colombia        3746    Male   
25812                     Canada       12396    Male   
6575        United Arab Emirates         104    Male   
2239                      Mexico       30715    Male   
154542                    Mexico       29398    Male   
23073                      Spain       21844    Male   
122887                    Mexico       20021  Female   
266315                     Italy       28859    Male   
187965                    Mexico       25385    Male   
82146                     Canada       15616  Female   
209539      United Arab Emirates        4085  Female   
81806                     Canada       12593  Female   
82441                     Canada       14107    Male   
247911                     Qatar        2641    Male   
152517                     Spain       12978  Female   
160658                     Italy       24322  Female   
140275            China-Shanghai        5065  Female   
266970                     Italy       27398  Female   
40980                    Romania         111  Female   
21112                      Spain       15312    Male   
162039                     Spain       24671    Male   
263548                     Spain       16464    Male   
94005                     Canada       15368  Female   
261519                     Italy       27383    Male   
22882                     Mexico       23772  Female   
59314                      Italy       25305    Male   
12025                      Italy       20429  Female   

        Out-of-School Study Time - Homework  \
236087                                  2.0   
247391                                  NaN   
7396                                    NaN   
8882                                    NaN   
56942                                   NaN   
201906                                  6.0   
203488                                  NaN   
39333                                   1.0   
104173                                  NaN   
23532                                   5.0   
21046                              

In [51]:
# Remove 'Mother - Highest Schooling' since it is redundant
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [52]:
# Check to see if correct values remain
mother_edu['Education - Mother'].value_counts()

<ISCED level 3>    25503
<ISCED level 5>    21484
<ISCED level 4>     5688
<ISCED level 6>     5533
<ISCED level 2>     5160
<ISCED level 1>     1347
<ISCED level 0>      635
Name: Education - Mother, dtype: int64

In [53]:
# Completed_edu is no longer needed since it duplicates the information in 'Education - Mother'
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [54]:
# Ensure the column has been dropped
mother_edu.sample(10)

Country  Student ID  Gender  \
246094  United Arab Emirates       11153    Male   
62966                 Mexico       33131  Female   
120952                Mexico       31205  Female   
216538               Finland        8796  Female   
160743                 Italy       25462  Female   
275538                 Qatar        5085  Female   
116617                Canada       14692    Male   
71947                  Spain       19904    Male   
88546                  Spain       14482    Male   
52173                  Spain       23357  Female   

        Out-of-School Study Time - Homework  \
246094                                  1.0   
62966                                   NaN   
120952                                  6.0   
216538                                 10.0   
160743                                 21.0   
275538                                  6.0   
116617                                  NaN   
71947                                   5.0   
88546                                   1.0   
52173                                  10.0   

        Out-of-School Study Time - Guided Homework  \
246094                                         NaN   
62966                                          NaN   
120952                                         1.0   
216538                                         1.0   
160743                                        21.0   
275538                                         3.0   
116617                                         NaN   
71947                                          3.0   
88546                                          0.0   
52173                                          1.0   

        Out-of-School Study Time - Personal Tutor  \
246094                                        2.0   
62966                                         NaN   
120952                                        0.0   
216538                                        0.0   
160743                                        0.0   
275538                                        2.0   
116617                                        NaN   
71947                                         3.0   
88546                                         0.0   
52173                                         0.0   

        Out-of-School Study Time - Commercial Company  \
246094                                            NaN   
62966                                             NaN   
120952                                            0.0   
216538                                            0.0   
160743                                            0.0   
275538                                            NaN   
116617                                            NaN   
71947                                             0.0   
88546                                             0.0   
52173                                             0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
246094                                     1.0                        315.0   
62966                                      NaN                          NaN   
120952                                     1.0                        225.0   
216538                                     1.0                        168.0   
160743                                     0.0                        165.0   
275538                                     5.0                        300.0   
116617                                     NaN                          NaN   
71947                                      0.0                        150.0   
88546                                      0.0                        240.0   
52173                                      1.0                        200.0   

        Learning Time - Test Language  Learning Time - Science  \
246094                          315.0                    360.0   
62966                             NaN                      NaN   
120952                          270.0                  

In [55]:
# Check if there are any duplicated values remaining
mother_edu['Student ID'].duplicated().sum()

63456

In [56]:
# mother_edu.replace(["NaN"], np.nan, inplace = True)
# mother_edu.dropna(inplace = True)

In [57]:
# Change datatypes into category, and put into order from lowest to highest
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['<ISCED level 6>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 0>'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)

# Remove duplicates and keep last, which is the highest level of education for the student's father
mother_edu_clean = mother_edu.drop_duplicates(subset=['Student ID'], keep='first')

In [58]:
# Ensure there are no duplicates
mother_edu_clean['Student ID'].duplicated().sum()

0

In [59]:
# Check the size of the dataframe 
print(pisa['STIDSTD'].nunique())
print(mother_edu_clean.shape)

33806
(33731, 17)


In [60]:
# Update pisa_clean with final variable from mother and father education
pisa_clean = mother_edu_clean.copy()

In [61]:
mother_edu_clean.head()

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
59124   Italy       28303  Female                                 12.0   
54941   Spain       16107  Female                                  NaN   
54932   Spain       16120    Male                                  6.0   
60083  Canada       21391    Male                                 12.0   
55113  Mexico       19312    Male                                  NaN   

       Out-of-School Study Time - Guided Homework  \
59124                                         2.0   
54941                                         NaN   
54932                                         5.0   
60083                                         5.0   
55113                                         NaN   

       Out-of-School Study Time - Personal Tutor  \
59124                                        2.0   
54941                                        NaN   
54932                                        6.0   
60083                                        9.0   
55113                                        NaN   

       Out-of-School Study Time - Commercial Company  \
59124                                            0.0   
54941                                            NaN   
54932                                            6.0   
60083                                            6.0   
55113                                            NaN   

       Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
59124                                     0.0                        180.0   
54941                                     NaN                          NaN   
54932                                     0.0                        180.0   
60083                                     2.0                        300.0   
55113                                     NaN                          NaN   

       Learning Time - Test Language  Learning Time - Science  \
59124                          240.0                    120.0   
54941                            NaN                      NaN   
54932                          180.0                    180.0   
60083                          375.0                    150.0   
55113                            NaN                      NaN   

       Average Math Score  Average Reading Score  Average Science Score  \
59124           528.28324              520.91252              522.88326   
54941           542.61568              566.02934              593.56582   
54932           592.70134              583.71758              652.49908   
60083           453.73888              434.23498              476.25886   
55113           368.13368              356.36616              325.94178   

       Average Total Score Education - Father Education - Mother  
59124           524.026340    <ISCED level 5>    <ISCED level 6>  
54941           567.403613    <ISCED level 5>    <ISCED level 6>  
54932           609.639333    <ISCED level 5>    <ISCED level 6>  
60083           454.744240    <ISCED level 4>    <ISCED level 6>  
55113           350.147207    <ISCED level 5>    <ISCED level 6>

### 4. Missing values in relevant columns 
('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [62]:
# Change string "NaN" to real nan value, and drop those values
pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

In [63]:
# Check to see if null values correctly dropped
pisa_clean

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
59124    Italy       28303  Female                                 12.0   
54932    Spain       16120    Male                                  6.0   
60083   Canada       21391    Male                                 12.0   
59844   Canada       21089  Female                                  8.0   
55467   Mexico       33705    Male                                  1.0   
52891   Mexico       23855    Male                                  6.0   
55261   Mexico       19937  Female                                 10.0   
59704   Canada       16670    Male                                  1.0   
60109   Canada       20456    Male                                  3.0   
59754   Canada       17706    Male                                  7.0   
60012   Brazil       18103  Female                                  5.0   
67212   Mexico       33178  Female                                  2.0   
59870   Canada       20652  Female                                 15.0   
60100   Canada       21241    Male                                  0.0   
59464    Italy       29438    Male                                 21.0   
54967    Spain       15832  Female                                 20.0   
59266    Italy       23518  Female                                  5.0   
60051   Canada       18877    Male                                  2.0   
59821   Canada       15554  Female                                  5.0   
59657   Canada       11518    Male                                 21.0   
52913   Mexico       23913    Male                                  6.0   
59254    Italy       23585    Male                                  5.0   
54982    Spain       12270  Female                                  7.0   
59975   Brazil       14613  Female                                  6.0   
59981   Brazil       16413  Female                                  3.0   
54978    Spain        7516  Female                                  7.0   
55615   Mexico       24781    Male                                  2.0   
59840   Canada       21181    Male                                  2.0   
59974   Canada       19868  Female                                  5.0   
59483    Italy       29924    Male                                  1.0   
...        ...         ...     ...                                  ...   
324905  Mexico       17506  Female                                  4.0   
324827  Mexico       21947    Male                                  2.0   
325062  Mexico       19471    Male                                  2.0   
325563  Mexico       19433  Female                                  3.0   
325646  Mexico       20912    Male                                  5.0   
325667  Mexico       19741  Female                                 16.0   
325861  Mexico       17174  Female                                  3.0   
324812  Mexico       21209  Female                                  2.0   
323644  Mexico       29825  Female                                  7.0   
323912  Mexico       31700    Male                                  4.0   
324083  Mexico       26583    Male                                  4.0   
324765  Mexico       22491  Female                                 14.0   
324478  Mexico       15477    Male                                  5.0   
324579  Mexico       10242    Male                                  1.0   
324616  Mexico       20470    Male                                 28.0   
325992  Mexico       33388    Male                                  8.0   
326823  Mexico       30408    Male                                  2.0   
326904  Mexico       26241  Female                                  2.0   
326781  Mexico       30320    Male                                 12.0   
326947  Mexico       26173  Female                                  2.0   
326987  Mexico       21819  Female                                  3.0   
327033  Mexico       22264  Fema

In [64]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(15167, 17)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [65]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning Time - Total'] = pisa_clean['Learning Time - Mathematics'] + pisa_clean['Learning Time - Test Language'] + pisa_clean['Learning Time - Science'] 
                                                                                                                 

In [66]:
list(pisa_clean)

['Country',
 'Student ID',
 'Gender',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Education - Father',
 'Education - Mother',
 'Out-of-School Study Time - Total',
 'Learning Time - Total']

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [67]:
pisa_clean

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
59124    Italy       28303  Female                                 12.0   
54932    Spain       16120    Male                                  6.0   
60083   Canada       21391    Male                                 12.0   
59844   Canada       21089  Female                                  8.0   
55467   Mexico       33705    Male                                  1.0   
52891   Mexico       23855    Male                                  6.0   
55261   Mexico       19937  Female                                 10.0   
59704   Canada       16670    Male                                  1.0   
60109   Canada       20456    Male                                  3.0   
59754   Canada       17706    Male                                  7.0   
60012   Brazil       18103  Female                                  5.0   
67212   Mexico       33178  Female                                  2.0   
59870   Canada       20652  Female                                 15.0   
60100   Canada       21241    Male                                  0.0   
59464    Italy       29438    Male                                 21.0   
54967    Spain       15832  Female                                 20.0   
59266    Italy       23518  Female                                  5.0   
60051   Canada       18877    Male                                  2.0   
59821   Canada       15554  Female                                  5.0   
59657   Canada       11518    Male                                 21.0   
52913   Mexico       23913    Male                                  6.0   
59254    Italy       23585    Male                                  5.0   
54982    Spain       12270  Female                                  7.0   
59975   Brazil       14613  Female                                  6.0   
59981   Brazil       16413  Female                                  3.0   
54978    Spain        7516  Female                                  7.0   
55615   Mexico       24781    Male                                  2.0   
59840   Canada       21181    Male                                  2.0   
59974   Canada       19868  Female                                  5.0   
59483    Italy       29924    Male                                  1.0   
...        ...         ...     ...                                  ...   
324905  Mexico       17506  Female                                  4.0   
324827  Mexico       21947    Male                                  2.0   
325062  Mexico       19471    Male                                  2.0   
325563  Mexico       19433  Female                                  3.0   
325646  Mexico       20912    Male                                  5.0   
325667  Mexico       19741  Female                                 16.0   
325861  Mexico       17174  Female                                  3.0   
324812  Mexico       21209  Female                                  2.0   
323644  Mexico       29825  Female                                  7.0   
323912  Mexico       31700    Male                                  4.0   
324083  Mexico       26583    Male                                  4.0   
324765  Mexico       22491  Female                                 14.0   
324478  Mexico       15477    Male                                  5.0   
324579  Mexico       10242    Male                                  1.0   
324616  Mexico       20470    Male                                 28.0   
325992  Mexico       33388    Male                                  8.0   
326823  Mexico       30408    Male                                  2.0   
326904  Mexico       26241  Female                                  2.0   
326781  Mexico       30320    Male                                 12.0   
326947  Mexico       26173  Female                                  2.0   
326987  Mexico       21819  Female                                  3.0   
327033  Mexico       22264  Fema

In [68]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15167 entries, 59124 to 326633
Data columns (total 19 columns):
Country                                          15167 non-null object
Student ID                                       15167 non-null int64
Gender                                           15167 non-null object
Out-of-School Study Time - Homework              15167 non-null float64
Out-of-School Study Time - Guided Homework       15167 non-null float64
Out-of-School Study Time - Personal Tutor        15167 non-null float64
Out-of-School Study Time - Commercial Company    15167 non-null float64
Out-of-School Study Time - With Parent           15167 non-null float64
Learning Time - Mathematics                      15167 non-null float64
Learning Time - Test Language                    15167 non-null float64
Learning Time - Science                          15167 non-null float64
Average Math Score                               15167 non-null float64
Average Reading Score     

In [69]:
pisa_clean.shape

(15167, 19)

In [70]:
pisa_clean.duplicated().sum()

0

In [71]:
pisa_clean['Student ID'].duplicated().sum()

0

In [72]:
pisa_clean.shape

(15167, 19)

In [73]:
pisa.shape

(485490, 636)

In [74]:
pisa_clean.head()

Country  Student ID  Gender  Out-of-School Study Time - Homework  \
59124   Italy       28303  Female                                 12.0   
54932   Spain       16120    Male                                  6.0   
60083  Canada       21391    Male                                 12.0   
59844  Canada       21089  Female                                  8.0   
55467  Mexico       33705    Male                                  1.0   

       Out-of-School Study Time - Guided Homework  \
59124                                         2.0   
54932                                         5.0   
60083                                         5.0   
59844                                         7.0   
55467                                         1.0   

       Out-of-School Study Time - Personal Tutor  \
59124                                        2.0   
54932                                        6.0   
60083                                        9.0   
59844                                        0.0   
55467                                        0.0   

       Out-of-School Study Time - Commercial Company  \
59124                                            0.0   
54932                                            6.0   
60083                                            6.0   
59844                                            0.0   
55467                                            1.0   

       Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
59124                                     0.0                        180.0   
54932                                     0.0                        180.0   
60083                                     2.0                        300.0   
59844                                     4.0                        450.0   
55467                                     0.0                        240.0   

       Learning Time - Test Language  Learning Time - Science  \
59124                          240.0                    120.0   
54932                          180.0                    180.0   
60083                          375.0                    150.0   
59844                          450.0                    450.0   
55467                          180.0                    300.0   

       Average Math Score  Average Reading Score  Average Science Score  \
59124           528.28324              520.91252              522.88326   
54932           592.70134              583.71758              652.49908   
60083           453.73888              434.23498              476.25886   
59844           669.27092              655.70692              678.14250   
55467           331.83520              272.16190              374.80420   

       Average Total Score Education - Father Education - Mother  \
59124           524.026340    <ISCED level 5>    <ISCED level 6>   
54932           609.639333    <ISCED level 5>    <ISCED level 6>   
60083           454.744240    <ISCED level 4>    <ISCED level 6>   
59844           667.706780    <ISCED level 5>    <ISCED level 6>   
55467           326.267100    <ISCED level 5>    <ISCED level 6>   

       Out-of-School Study Time - Total  Learning Time - Total  
59124                              16.0                  540.0  
54932                              23.0                  540.0  
60083                              34.0                  825.0  
59844                              19.0                 1350.0  
55467                               3.0                  720.0

<a id='sav'></a>
## Store

In [75]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References